carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/ParticipantGroupStatusService.kt

```
@@ -0,0 +1,102 @@
+                    id = group.id,
+                    participants = participants,
+                    assignedParticipantRoles = group.roleAssignments,
+                    name = group.name
+                )
+            }
+
+        /**
+         * Get the [ParticipantGroupStatus] for multiple participant [groups].
+         * Deployed groups request their deployment status using [getDeploymentStatusList] in a single aggregated call.
+         *
+         * @throws IllegalArgumentException when the returned deployment status IDs don't match the requested
+         * deployed group IDs.
+         */
+        suspend fun forGroups(
+            groups: Collection<StagedParticipantGroup>,
+            getParticipants: (StagedParticipantGroup) -> Set<Participant>,
+            getDeploymentStatusList: suspend (Set<UUID>) -> List<StudyDeploymentStatus>
+        ): List<ParticipantGroupStatus>
+        {
+            val (deployedGroups, stagedGroups) = groups.partition { it.isDeployed }
+
+            // Get status for staged groups.
+            val stagedStatuses = stagedGroups.map { group ->
+                ParticipantGroupStatus.Staged(
+                    id = group.id,
+                    participants = getParticipants( group ),
+                    assignedParticipantRoles = group.roleAssignments,
+                    name = group.name
+                )
+            }
+
+            // Get deployment statuses for deployed groups.
+            val deployedIds = deployedGroups.map { it.id }.toSet()
+            val deploymentStatuses =
+                if ( deployedIds.isEmpty() ) emptyList()
+                else getDeploymentStatusList( deployedIds )
+
+            val deploymentStatusIds = deploymentStatuses.map { it.studyDeploymentId }.toSet()
+            require( deploymentStatusIds == deployedIds )
+            {
+                "Deployment status IDs \"$deploymentStatusIds\" don't match requested group IDs \"$deployedIds\"."
+            }
```

There is a change in strictness, `Recruitment.getParticipantGroupStatus` only require each deployment status belongs to current recruitment, now we also check if there are missing/exceeding deployment status compare to participant groups.

---

When I refactored this I noticed this, and I thought it best to be accepting (so if there are "exceeding" deployment status returned, not fail, but of course we need the status for those requested, and we want to fail early if anything is missing.